lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelectionSpec.hs

```diff
@@ -254,6 +270,120 @@ prop_performSelection_onSuccess_hasSuitableCollateral cs _ps selection =
     suitableForCollateral :: (TxIn, TxOut) -> Bool
     suitableForCollateral = isJust . view #utxoSuitableForCollateral cs
 
+--------------------------------------------------------------------------------
+-- Construction of balance constraints and parameters
+--------------------------------------------------------------------------------
+
+-- Tests that function 'toBalanceConstraintsParams' applies the correct
+-- transformation to the 'computeMinimumCost' function.
+--
+prop_toBalanceConstraintsParams_computeMinimumCost
+    :: MockSelectionConstraints
+    -> SelectionParams
+    -> SelectionSkeleton
+    -> Property
+prop_toBalanceConstraintsParams_computeMinimumCost
+    mockConstraints params skeleton =
+        checkCoverage $
+        cover 10 (selectionCollateralRequired params)
+            "collateral required: yes" $
+        cover 10 (not (selectionCollateralRequired params))
+            "collateral required: no" $
+        cover 10 (costOriginal < costAdjusted)
+            "cost (original) < cost (adjusted)" $
+        report costOriginal
+            "cost (original)" $
+        report costAdjusted
+            "cost (adjusted)" $
+        if selectionCollateralRequired params
+        then
+            conjoin
+                [ costOriginal <= costAdjusted
+                -- Here we apply a transformation that is the *inverse* of
+                -- the transformation within 'toBalanceConstraintsParams':
+                , costOriginal ==
+                    ( computeMinimumCostAdjusted
+                    . over #skeletonInputCount
+                        (subtract maximumCollateralInputCount)
+                    $ skeleton
+                    )
+                ]
+        else
+            costOriginal === costAdjusted
+  where
+    constraints :: SelectionConstraints
+    constraints = unMockSelectionConstraints mockConstraints
+
+    maximumCollateralInputCount :: Int
+    maximumCollateralInputCount = constraints ^. #maximumCollateralInputCount
+
+    computeMinimumCostOriginal :: SelectionSkeleton -> Coin
+    computeMinimumCostOriginal = constraints ^. #computeMinimumCost
+
+    computeMinimumCostAdjusted :: SelectionSkeleton -> Coin
+    computeMinimumCostAdjusted =
+        toBalanceConstraintsParams (constraints, params)
+            & fst & view #computeMinimumCost
+
+    costOriginal :: Coin
+    costOriginal = computeMinimumCostOriginal skeleton
```

Comment:
Looking at this again it occurred to me: isn't it strange that the outer constraints contain the unadjusted and wrong minimum cost? If someone were to call the outer `computeMinimumCost` they would get the wrong results?